browser_patches/firefox-beta/juggler/components/juggler.js

```diff
@@ -17,115 +19,111 @@ const Ci = Components.interfaces;
 
 const FRAME_SCRIPT = "chrome://juggler/content/content/main.js";
 
-// Command Line Handler
-function CommandLineHandler() {
-};
+class Juggler {
+  get classDescription() { return "Sample command-line handler"; }
+  get classID() { return Components.ID('{f7a74a33-e2ab-422d-b022-4fb213dd2639}'); }
+  get contractID() { return "@mozilla.org/remote/juggler;1" }
+  get QueryInterface() {
+    return ChromeUtils.generateQI([ Ci.nsICommandLineHandler, Ci.nsIObserver ]);
+  }
+  get helpInfo() {
+    return "  --juggler            Enable Juggler automation\n";
+  }
 
-CommandLineHandler.prototype = {
-  classDescription: "Sample command-line handler",
-  classID: Components.ID('{f7a74a33-e2ab-422d-b022-4fb213dd2639}'),
-  contractID: "@mozilla.org/remote/juggler;1",
-  _xpcom_categories: [{
-    category: "command-line-handler",
-    entry: "m-juggler"
-  }],
+  handle(cmdLine) {
+    // flag has to be consumed in nsICommandLineHandler:handle
+    // to avoid issues on macos. See Marionette.jsm::handle() for more details.
+    // TODO: remove after Bug 1724251 is fixed.
+    cmdLine.handleFlag("juggler-pipe", false);
+  }
 
-  /* nsICommandLineHandler */
-  handle: async function(cmdLine) {
-    const jugglerFlag = cmdLine.handleFlagWithParam("juggler", false);
-    const jugglerPipeFlag = cmdLine.handleFlag("juggler-pipe", false);
-    if (!jugglerPipeFlag && (!jugglerFlag || isNaN(jugglerFlag)))
-      return;
-    const silent = cmdLine.preventDefault;
-    if (silent)
-      Services.startup.enterLastWindowClosingSurvivalArea();
+  async observe(subject, topic) {
+    switch (topic) {
+      case "profile-after-change":
+        Services.obs.addObserver(this, "command-line-startup");
+        break;
+      case "command-line-startup":
+        Services.obs.removeObserver(this, topic);
+        const cmdLine = subject;
+        const jugglerPipeFlag = cmdLine.handleFlag('juggler-pipe', false);
+        if (!jugglerPipeFlag)
+          return;
+        Services.startup.enterLastWindowClosingSurvivalArea();
+        Services.obs.addObserver(this, "final-ui-startup");
+        Services.obs.addObserver(this, "browser-idle-startup-tasks-finished");
+        Services.obs.addObserver(this, "mail-idle-startup-tasks-finished");
+        Services.obs.addObserver(this, "quit-application");
+        break;
+      // Used to wait until the initial application window has been opened.
+      case "final-ui-startup":
```

Comment:
Isn't this too late? Why can't we do this in `command-line-startup` instead?

Reply:
should be fine since this is what both remote and marionette do. 

Earlier stage is scary since e.g. preferences are not ready yet!